src/Microsoft.AspNet.Server.Kestrel/Infrastructure/MemoryPoolIterator2.cs

```diff
@@ -295,6 +298,45 @@ public int Seek(int char0, int char1)
             }
         }
 
+        /// <summary>
+        /// Save the data at the current location then move to the next available space.
+        /// </summary>
+        /// <param name="data">The byte to be saved.</param>
+        /// <returns>0 if the operation successes. -1 if can't find available space.</returns>
+        public int Put(byte data)
+        {
+            if (_block == null)
+            {
+                return -1;
+            }
+            else if (_index < _block.End)
+            {
+                _block.Array[_index++] = data;
+                return 0;
+            }
+
+            var block = _block;
+            var index = _index;
+            while (true)
+            {
+                if (index < block.End)
+                {
+                    _block = block;
+                    _index = _index + 1;
+                    block.Array[index] = data;
```

Comment:
I think you should exit the method here. You only when block.Next == null in this loop.

Add a test for this.